utils/json_profiles_merger_lib.py

```diff
@@ -0,0 +1,173 @@
+  """Returns the median of the input list.
+
+  Args:
+    lst: the input list.
+
+  Returns:
+    The median of the list, or None if the list is empty/None.
+  """
+  sorted_lst = sorted(lst)
+  length = len(sorted_lst)
+  if length % 2:
+    return sorted_lst[length // 2]
+  return (sorted_lst[length // 2 - 1] + sorted_lst[length // 2]) / 2
+
+
+def _write_to_csv(
+    bazel_source, project_source, project_commit, event_list, output_csv_path):
+  """Writes the event_list to output_csv_path.
+
+  event_list format:
+  [{'cat': <string>, 'name': <string>, 'dur': <int>}, ...]
+
+  Args:
+    bazel_source: the bazel commit or path to the bazel binary from which these
+      JSON profiles were collected.
+    project_source: the project on which the runs that generated these JSON
+      projects were performed.
+    project_commit: the project commit on which the Bazel runs were performed.
+    event_list: the list of events, aggregated from the JSON profiles.
+    output_csv_path: a path to the output CSV file.
+  """
+  with open(output_csv_path, 'w') as csv_file:
+    csv_writer = csv.writer(csv_file)
+    csv_writer.writerow(
+        ['bazel_source', 'project_source', 'project_commit',
+         'cat', 'name', 'dur'])
+
+    for event in event_list:
+      csv_writer.writerow(
+          [bazel_source, project_source, project_commit,
+           event['cat'], event['name'], event['dur']])
+
+
+def _accumulate_event_duration(event_list, accum_dict):
+  """Fill up accum_dict by accummulating durations of each event.
+
+  Also create the entries for each phase by subtracting the build phase markers'
+  ts attribute.
+
+  Args:
+    event_list: the list of event objects.
+    accum_dict: the dict to be filled up with a mapping of the following format:
+      { <name>: { name: ..., cat: ..., dur_list: [...]}, ...}
+  """
+  # A list of tuples of the form (marker, occurrence time in micro s)
+  build_markers_ts_pairs = []
+
+  # Only collect events with a duration.
+  # Special case: markers that indicates beginning/end of execution.
+  for event in event_list:
+    if 'cat' in event and event['cat'] == 'build phase marker':
+      build_markers_ts_pairs.append((event['name'], event['ts']))
+
+    if 'dur' not in event:
+      continue
+
+    if event['name'] not in accum_dict:
+      accum_dict[event['name']] = {
+          'name': event['name'],
+          'cat': event['cat'],
+          'dur_list': []
+      }
+    accum_dict[event['name']]['dur_list'].append(event['dur'])
+
+  # Fill in the markers.
+  for i, marker_ts_pair in enumerate(build_markers_ts_pairs[:-1]):
+    marker, ts = marker_ts_pair
+    _, next_ts = build_markers_ts_pairs[i + 1]
+
+    if marker not in accum_dict:
+      accum_dict[marker] = {
+          'name': marker,
+          'cat': 'build phase marker',
+          'dur_list': []
+      }
+    current_phase_duration_millis = (next_ts - ts) / 1000
+    accum_dict[marker]['dur_list'].append(current_phase_duration_millis)
+
```

Comment:
what about the last phase? extend it till the end?

Reply:
Done